Kernel: Implemented shared memory.

Subv opened the pull request:
This PR corrects the implementation of Shared Memory thorough the codebase.
SharedMemories created by HLE processes are backed by memory located in the BASE linear heap region (Unless stated otherwise by the call to Create), as per the real kernel.

With this we can get rid of some very nasty hacks, like the APT Shared Font being located at a fixed address. This PR implements a function to automatically relocate the dumped shared font from the 0x18000000 address to whatever address the allocator places it at. This fixes an issue where we were placing the shared font in the middle of the APPLICATION memory region (Possibly breaking games that use that part of the region).

All error codes were verified by reverse engineering unless otherwise stated.
There are some sanity checks that can not be performed unless we give each HLE service its own Process, but that's a discussion for another time.

Closes #1603
Closes #1547
Closes #1591
Closes #1318
Closes #1596
Closes #1497

wwylele commented:
I am not sure if I did something wrong, but with this PR, all of my games can't boot and spam with:
`HW.Memory <Error> core\memory.cpp:Memory::Read:137: unmapped Read8 @ 0x10002001`

mailwl commented:
Kid Icarus Uprising - Won't boot #1603, the real candidate for testing this PR throws error:
> [  21.386077] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:108: Fatal error!
> [  21.391196] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:116: PID: 0x00000000_0x00000000
> [  21.391607] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:117: REV: 44677
> [  21.391972] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:118: AID: 0x00000000_0x00000000
> [  21.392318] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:119: ADR: 0x00185C64
> [  21.392640] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:121: RSL: 0xE0E01BEE
> [  21.393029] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:122:   Level: 28
> [  21.393330] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:123:   Summary: 7
> [  21.393605] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:124:   Module: 6
> [  21.397221] Service.ERR <Critical> core\hle\service\err_f.cpp:ERR_F::ThrowFatalError:125:   Desc: 1006

Subv commented:
There was an operator precedence error in the code, i've now updated it, could you please test again?

yuriks commented on src/core/hle/kernel/shared_memory.h:
```diff
@@ -48,6 +48,12 @@ class SharedMemory final : public Object {
+     * Converts the specified MemoryPermission into the equivalent VMAPermission.
+     * @param permission The MemoryPermission to convert.
+     */
+    static VMAPermission ConvertPermissions(MemoryPermission permission);
```
This should just be a free local function in shared_memory.cpp

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -85,15 +109,29 @@ ResultCode SharedMemory::Map(SharedPtr<Process> target_process, VAddr address, M
+    if (static_cast<u32>(permission) & static_cast<u32>(MemoryPermission::Write))
+        vma_permission |= static_cast<u32>(VMAPermission::Write);
+    if (static_cast<u32>(permission) & static_cast<u32>(MemoryPermission::Execute))
+        vma_permission |= static_cast<u32>(VMAPermission::Execute);
```
Except for DONT_CARE, VMAPermission purposefully matches the bits of SVC interface type, so you can just mask ensure there are no inappropriate bits set and cast.

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
+        MemoryRegionInfo* memory_region = GetMemoryRegion(region);
+        auto& linheap_memory = memory_region->linear_heap_memory;
+
+        ASSERT_MSG(linheap_memory->size() + size <= memory_region->size, "Not enough space in region to allocate shared memory!");
```
Shouldn't this return an error code instead?

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
+    // can not map it in its own address space unless it was created with addr=0, result 0xD900182C.
 
-    this->base_address = address;
+    if (address) {
```
address != 0

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
+
+        shared_memory->backing_block = linheap_memory;
+        shared_memory->backing_block_offset = linheap_memory->size();
+        // Allocate some memory from the bottom of the linear heap for this region.
```
"from the end"* (less ambiguous)

yuriks commented on src/core/hle/service/apt/apt.cpp:
```diff
@@ -71,26 +126,88 @@ void Initialize(Service::Interface* self) {
-    }
+    // The shared font has to be relocated to the new address before being passed to the application.
+    VAddr target_address = Memory::PhysicalToVirtualAddress(shared_font_mem->linear_heap_phys_address);
+    // The shared font dumped by 3dsutils (https://github.com/citra-emu/3dsutils) uses this address as base,
```
This isn't true if you dump it on a n3DS. Is there a way to detect the address instead? (Ideally we should un-relocate while dumping, or you can use some pointer that's near the beginning of the file to guess the offset.)

yuriks commented on src/core/hle/service/apt/apt.cpp:
```diff
@@ -433,14 +550,12 @@ void Init() {
         using Kernel::MemoryPermission;
-        shared_font_mem = Kernel::SharedMemory::Create(3 * 1024 * 1024, // 3MB
-                MemoryPermission::ReadWrite, MemoryPermission::Read, "APT_U:shared_font_mem");
+        shared_font_mem = Kernel::SharedMemory::Create(3272 * 1024, // 3272 KB
```
IMO just use 0x332000 here. Makes it more obvious that it's inherently 4k-aligned too (which isn't obvious like this)

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
+        shared_memory->backing_block = linheap_memory;
+        shared_memory->backing_block_offset = linheap_memory->size();
+        // Allocate some memory from the bottom of the linear heap for this region.
+        linheap_memory->insert(linheap_memory->end(), size, 0);
```
Needs this to avoid dangling pointers in the page table:
```
        if (Kernel::g_current_process != nullptr) {
            Kernel::g_current_process->vm_manager.RefreshMemoryBlockMappings(linheap_memory.get());
        }
```

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
+        MemoryPermission other_permissions, VAddr address, MemoryRegion region, std::string name) {
     SharedPtr<SharedMemory> shared_memory(new SharedMemory);
 
+    shared_memory->owner_process = Kernel::g_current_process;
```
This should be passed in as a parameter instead.

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
+        shared_memory->linear_heap_phys_address = Memory::FCRAM_PADDR + memory_region->base + shared_memory->backing_block_offset;
+    } else {
+        // The memory is already available and mapped in the owner process.
+        auto vma = shared_memory->owner_process->vm_manager.FindVMA(address)->second;
```
Make a reference to the vm_manager since it's used a bunch of times.

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
 }
 
-ResultCode SharedMemory::Map(VAddr address, MemoryPermission permissions,
+ResultCode SharedMemory::Map(SharedPtr<Process> target_process, VAddr address, MemoryPermission permissions,
```
This isn't stored anywhere (aka, doesn't need to preserve lifetime) so it can just be a regular pointer

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
+}
 
-    return RESULT_SUCCESS;
+ResultCode SharedMemory::Unmap(SharedPtr<Process> target_process, VAddr address) {
```
Likewise, regular pointer

yuriks commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -15,92 +16,124 @@ SharedMemory::SharedMemory() {}
 
-    return RESULT_SUCCESS;
+ResultCode SharedMemory::Unmap(SharedPtr<Process> target_process, VAddr address) {
+    return target_process->vm_manager.UnmapRange(address, size);
```
Does the 3DS kernel check that it's indeed a shared memory that's mapped here? What happens if you try to unmap a random address? This isn't checked at all in the svc handler either.

yuriks commented on src/core/hle/kernel/shared_memory.h:
```diff
@@ -66,10 +77,16 @@ class SharedMemory final : public Object {
+    /// Process that created this shared memory block.
+    SharedPtr<Process> owner_process;
+    /// Address of shared memory block in the owner process if specified.
     VAddr base_address;
```
This doesn't seem to be used anymore except for checking if it is equal to 0. Do you still need to keep this? (It also doesn't really make sense outside of the creating process.)

Subv replied:
For now it's basically acting as a linear_heap_allocation variable, but i consider it good to keep it around even if just for information's sake

yuriks commented on src/core/hle/service/apt/apt.cpp:
```diff
@@ -22,23 +22,78 @@
-// be relocated. This might be fixed by dumping the shared font @ address 0x00000000 and then
-// correctly mapping it in Citra, however we still do not understand how the mapping is determined.
-static const VAddr SHARED_FONT_VADDR = 0x18000000;
+/// BCFNT Shared Font file structures
```
I think you should move this and the relocation to separate bcfnt file(s) in the apt subdir.

JayFoxRox replied:
Agreed, I'm still hoping someone will make it possible to run without the font dump by creating a font file in memory at runtime (Rendered from ttf or something).

yuriks commented on src/core/hle/service/apt/apt.cpp:
```diff
@@ -71,26 +126,88 @@ void Initialize(Service::Interface* self) {
+    cmd_buff[0] = IPC::MakeHeader(0x44, 2, 2);
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    // Since the SharedMemory interface doesn't provide the address at which the memory was allocated,
+    // the APT service calculates this address by scanning the entire address space (using svcQueryMemory)
```
"the *real* APT service", just to clarify

wwylele commented:
Test with some random games, still some problems:
**Super Mario 3D Land** don't boot, and `ThrowFatelError`.
**Kirby Triple Deluxe** and **Pokemon Mystery Dungeon: Gate to Infinity** boot, but don't receive any input.
Some other games work fine.

yuriks commented:
@wwylele You need to add these lines to fix a bug: https://github.com/citra-emu/citra/pull/1689#discussion_r60173069

wwylele commented:
@yuriks Thanks. Kirby and Pokemon are OK now, but Mario still throws fatal error.

Subv commented:
Addressed some comments, also fixed the problem with Mario 3D Land (Logic error in CreateMemoryBlock) @wwylele

wwylele commented:
@Subv sorry for late reply. Mario 3D Land works fine now and no more game broken.

Subv commented:
Updated again and addressed most comments, some of the sanity checks (mostly edge cases) when freeing memory are not implemented because I don't currently have a way to verify their behavior.

jroweboy commented on src/core/hle/service/apt/apt.cpp:
```diff
@@ -71,26 +63,30 @@ void Initialize(Service::Interface* self) {
 }
 
+
+
```
Minor style nitpick: extra blank spaces

JayFoxRox commented on src/core/hle/service/apt/bcfnt/bcfnt.cpp:
```diff
@@ -0,0 +1,71 @@
+    for (unsigned block = 0; block < cfnt.num_blocks; ++block) {
+
+        u32 section_size = 0;
+        if (memcmp(data, "FINF", 4) == 0) {
```
I'm not an expert on char encoding, but can we assure that this will work on all machines?
I'd expect each symbol to use 2 bytes in some encodings and that could mean trouble.
Maybe create a const for the magic you are actually looking for?

wwylele commented:
software keyboard and mii selector are broken. Super Mario 3D land for example when creating new file:
```
[  24.349965] Service <Error> core\hle\service\service.cpp:Service::Interface::S
yncRequest:69: unknown / unimplemented function 'ReplySleepQuery': port=APT:U, c
md_buff[1]=0x300, cmd_buff[2]=0x66630000
[  24.351432] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::A
ppletUtility:300: (STUBBED) called command=0x00000004, buffer1_size=0x00000001,
buffer2_size=0x00000001, buffer1_addr=0x0FFFFC9C, buffer2_addr=0x0FFFFC98
[  24.352993] Service <Error> core\hle\service\service.cpp:Service::Interface::S
yncRequest:69: unknown / unimplemented function 'ReplySleepQuery': port=APT:U, c
md_buff[1]=0x300, cmd_buff[2]=0x66630000
[  24.354790] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::G
etAppletInfo:393: (stubbed) called appid=1026
[  24.356181] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::A
ppletUtility:300: (STUBBED) called command=0x00000004, buffer1_size=0x00000001,
buffer2_size=0x00000001, buffer1_addr=0x0FFFFB0C, buffer2_addr=0x0FFFFB08
[  24.357373] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::I
sRegistered:150: (STUBBED) called app_id=0x00000402
[  24.358328] Service <Error> core\hle\service\service.cpp:Service::Interface::S
yncRequest:69: unknown / unimplemented function 'SaveVramSysArea': port=gsp::Gpu

[  24.359557] Service.GSP <Warning> core\hle\service\gsp_gpu.cpp:GSP_GPU::Import
DisplayCaptureInfo:630: called
[  24.360477] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::I
sRegistered:150: (STUBBED) called app_id=0x00000402
[  24.361445] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::S
endParameter:192: (STUBBED) called src_app_id=0x00000300, dst_app_id=0x00000402,
 signal_type=0x00000002,buffer_size=0x00000020, value=0x00000000, handle=0x00000
000, size=0x00080002, in_param_buffer_ptr=0x003F9400
[  24.363327] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::G
lanceParameter:238: called app_id=0x00000300, buffer_size=0x00001000
[  24.364313] Service.APT <Warning> core\hle\service\apt\apt.cpp:Service::APT::R
eceiveParameter:215: called app_id=0x00000300, buffer_size=0x00000000
```
(Till here is the same as it in master)
```
[ 113.753902] HW.Memory <Error> core\memory.cpp:Memory::Write:222: unmapped Writ
e32 0xFE60FE60 @ 0x10004000
[ 113.962197] HW.Memory <Error> core\memory.cpp:Memory::Write:222: unmapped Writ
e32 0xFE60FE60 @ 0x10004004
[ 114.169670] HW.Memory <Error> core\memory.cpp:Memory::Write:222: unmapped Writ
e32 0xFE60FE60 @ 0x10004008
...
```

wwylele commented on src/core/hle/kernel/shared_memory.cpp:
```diff
@@ -7,100 +7,148 @@
+    MemoryPermission own_other_permissions = target_process == owner_process ? this->permissions : this->other_permissions;
+
+    // Automatically allocated memory blocks can only be mapped with other_permissions = DontCare
+    if (base_address == 0 && other_permissions != MemoryPermission::DontCare) {
```
When Super Mario 3D Land calling mii selector, it wants to map the mii selector shared memory (`base_address=0`) with `other_permissions = MemoryPermission::ReadWrite`

Subv replied:
Interesting, i'll take another look at the kernel code later today

Leo626 commented:
Fixed the text in **Sadame**.
![citra-qt_2016-05-01_20-02-44](https://cloud.githubusercontent.com/assets/17358613/14945145/206d675e-0fd8-11e6-9ef0-a75a57f9800d.png)
![citra-qt_2016-05-01_20-01-11](https://cloud.githubusercontent.com/assets/17358613/14945146/2786493e-0fd8-11e6-8097-7494defd7807.png)

JohnGodgames commented:
**With this PR:**
- Zoo Keeper 3D boots again
- Real Heroes Firefighter 3D boots again (show loading screen)

Subv commented:
@wwylele I have a fix for the applet issue, I'll update this PR as soon as #1761 is merged

Subv commented:
@wwylele Rebased and updated, the problem with applets should be fixed now.

wwylele commented:
@Subv Nice, it works well at software keyboard and mii selector now. Didn't find anything broken more

tfarley commented on src/core/hle/service/csnd_snd.cpp:
```diff
@@ -41,9 +42,9 @@ static Kernel::SharedPtr<Kernel::Mutex> mutex = nullptr;
-            Kernel::MemoryPermission::ReadWrite,
-            Kernel::MemoryPermission::ReadWrite, "CSNDSharedMem");
+    u32 size = Common::AlignUp(cmd_buff[1], Memory::PAGE_SIZE);
+    shared_memory = Kernel::SharedMemory::Create(nullptr, size,
```
Text wrap like gsp_gpu.cpp/hid.cpp/ir.cpp

JayFoxRox commented on src/core/hle/function_wrappers.h:
```diff
@@ -170,7 +170,7 @@ template<ResultCode func(s64*, u32, s32)> void Wrap() {
 template<ResultCode func(u32*, u32, u32, u32, u32)> void Wrap() {
     u32 param_1 = 0;
-    u32 retval = func(&param_1, PARAM(1), PARAM(2), PARAM(3), PARAM(4)).raw;
+    u32 retval = func(&param_1, PARAM(1), PARAM(2), PARAM(3), PARAM(0)).raw;
```
This feels weird, there should be a comment why this has to be like that or at least a comment that this is intentional.

Subv replied:
This file is pretty much a "write and forget" thing, it's awful and full of little things that may break without anyone noticing (Like this function). /endrant

The reason why the arguments are ordered like that is not known, that's just what Nintendo's Kernel Calling Convention looks like (It's pretty inconsistent). I don't really see the value of placing a comment here.

hrydgard replied:
Since it's rather unexpected and looks like a typo if you're not aware, I would agree with JayFoxRox that it does merit a (short) comment.

Subv replied:
I've added a comment about this in the code.

Subv commented:
Squashed, this one is good to go.

bunnei commented:
LGTM

KillzXGaming commented:
Seems this caused regression and broke Animal Crossing New Leaf. Can anyone else confirm this? I get these errors once booting and black screen. 
  21.516452] HW.Memory <Error> core\memory.cpp:Memory::Read:181: unmapped Read8 @ 0x10000001

JohnGodgames commented:
My Animal Crossing New Leaf (US) works fine.
